include/p2300/tag_invoke.hpp

```diff
@@ -28,9 +28,71 @@
 
 // ----------------------------------------------------------------------------
 // [func.tag_invoke]
+#include <type_traits>
 
-namespace nstd {
-}
+namespace std {
+    template <auto& _Tag>
+    using tag_t = decay_t<decltype(_Tag)>;
+    namespace _Tag_invoke {
+        void tag_invoke();
+
+        template <class _Ty>
+        concept _Has_class_or_enum_type = __is_class(remove_reference_t<_Ty>)
+            || __is_enum(remove_reference_t<_Ty>)
+            || __is_union(remove_reference_t<_Ty>);
+
+        template <class _Tag, class... _Args>
+        concept _Tag_invocable_no_adl = _Has_class_or_enum_type<_Tag> && requires {
+            {tag_invoke(std::declval<_Tag>(), std::declval<_Args>()...)};
+        };
+
+        template <class _Tag, class... _Args>
+        concept _Nothrow_tag_invocable_no_adl = _Has_class_or_enum_type<_Tag> && requires {
+            {tag_invoke(std::declval<_Tag>(), std::declval<_Args>()...)} noexcept;
+        };
+
+        template <class _Tag, class _Arg>
+        concept _Tag_observable_no_adl = _Has_class_or_enum_type<_Tag> && requires {
+            {tag_invoke(std::declval<_Tag>(), std::declval<add_const_t<_Arg>>())} noexcept;
+        };
```

Comment:
The name `_Tag_observable` isn't mention in P2300: is this concept used for the implementation of other components? The defintion/name seems to imply it is used for observer/accessor operations to verify that a `const`-qualifed second argument would also work.

Reply:
Yeah, I should have maybe named it `_Tag_querriable` but that is a horrible word.

In a nutshell it should cover all querries that take the CPO by const reference and are noexcept. We have a few of them and centralizing them reduces the risk of errors

Reply:
OK. I wasn't worried about the [internal] name. I just wondered why this is here as the spec doesn't say anything about it and I didn't see it used [yet].